File: backend/src/main/java/com/fptgang/backend/model/Order.java
package com.fptgang.backend.model;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;

@Entity
@Table(name = "`order`") // Escape the reserved keyword "order"
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Order {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long orderId;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(nullable = false, name = "account_id")
    private Account account;

    @OneToMany(mappedBy = "order", fetch = FetchType.LAZY, cascade = CascadeType.ALL, orphanRemoval = true)
    @ToString.Exclude
    private List<OrderStatusHistory> orderStatusHistories; // OrderStatusHistory>

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private OrderStatusHistory.State latestStatus;

    @OneToMany(mappedBy = "order", fetch = FetchType.LAZY, cascade = CascadeType.ALL, orphanRemoval = true)
    @ToString.Exclude
    private List<OrderDetail> orderDetails;

    @OneToOne(mappedBy = "order", fetch = FetchType.LAZY, cascade = CascadeType.ALL, orphanRemoval = true)
    private Transaction transaction;

    @OneToOne(mappedBy = "order", fetch = FetchType.LAZY)
    private Voucher voucher;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "shipping_info_id")
    private ShippingInfo shippingInfo;

    @CreationTimestamp
    private LocalDateTime createdAt;

    @UpdateTimestamp
    private LocalDateTime updatedAt;

    @Column(nullable = false, precision = 10, scale = 2)
    private BigDecimal subTotal;

    @Column(nullable = false, precision = 10, scale = 2)
    private BigDecimal finalTotal;

}

File: gendb/src/engine/model/Order.ts
import { OrderDetail } from "./OrderDetail";
import {Voucher} from "./Voucher";
import {Account} from "./Account";
import {OrderState} from "./OrderStatusHistory";

export class Order {
    order_id: number;
    final_total: number;
    created_at: Date;
    sub_total: number;
    latest_status: OrderState;
    updated_at: Date;
    account_id: number;
    shipping_info_id: number | null;

    // convenient fields, do not dump
    details?: OrderDetail[];
    voucher: Voucher | null = null;
    account: Account | null = null;

    constructor(init: Partial<Order>) {
        this.order_id = init.order_id || 0;
        this.final_total = init.final_total || 0;
        this.created_at = init.created_at || new Date();
        this.sub_total = init.sub_total || 0;
        this.updated_at = init.updated_at || new Date();
        this.latest_status = init.latest_status || OrderState.CREATED;
        this.account_id = init.account_id || 0;
        this.shipping_info_id = init.shipping_info_id || null;
        this.details = init.details || [];
        this.voucher = init.voucher || null;
        this.account = init.account || null;
    }

    static dump(orders: Order[]): string {
        if (orders.length === 0) return '';

        const fields = [
            'order_id',
            'final_total',
            'created_at',
            'sub_total',
            'latest_status',
            'updated_at',
            'account_id',
            'shipping_info_id'
        ];

        const values = orders.map(order => {
            return `(${[
                order.order_id,
                order.final_total.toFixed(2),
                `'${order.created_at.toISOString().slice(0, 19).replace('T', ' ')}.000000'`,
                order.sub_total.toFixed(2),
                `'${order.latest_status}'`,
                `'${order.updated_at.toISOString().slice(0, 19).replace('T', ' ')}.000000'`,
                order.account_id,
                order.shipping_info_id === null ? 'NULL' : order.shipping_info_id
            ].join(',')})`
        }).join(',\n');

        return `INSERT INTO \`order\` (${fields.map(f => `\`${f}\``).join(',')}) VALUES\n${values};`;
    }
}